Avoid get_notes() in Notes::possibly_delete_survey_notes()

jacob-sewell opened the pull request:
Fixes #7987

The methods `possibly_delete_survey_notes()` and `possibly_delete_marketing_notes()` in `Automattic\WooCommerce\Admin\Notes\Notes` used to call `get_notes()` in the data store, which could get expensive.

This PR adds a method to the data store, `get_notes_with_type()`, which works the same as the existing `get_notes_with_name()` except for the parameter it searches on. Importantly, it does not use the `'woocommerce_note_where_clauses'` filter.

`Notes::possibly_delete_survey_notes()` and `Notes::possibly_delete_marketing_notes()` have been rewritten to use this new method instead of `get_notes()`.

### Detailed test instructions:

Survey note deletion is triggered in the [admin_init](https://developer.wordpress.org/reference/hooks/admin_init/) hook, and marketing note deletion is triggered when the woocommerce_show_marketplace_suggestions option is set to 'no'.
<!--
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note using the following CLI command `npm run changelogger -- add` and commit changes. --->

jacob-sewell committed:
Query for admin survey note_ids directly in Notes::possibly_delete_survey_notes().

jacob-sewell commented:
This is going to remain in draft until we settle on how we want to tackle PR #8202 , because this one's probably going the same way.

joshuatf commented on src/Notes/Notes.php:
```diff
@@ -272,16 +272,19 @@ public static function possibly_delete_marketing_notes( $old_value, $value ) {
+		global $wpdb;
+		$note_ids = $wpdb->get_col(
+			$wpdb->prepare(
+				"SELECT `note_id` FROM {$wpdb->prefix}wc_admin_notes WHERE `type` = %s AND `status` = %s",
```
I think this is confirming your suspicions about circumventing the `get_notes()` call and expectations around filters.

I would argue that this is a third party issue and not one we should be manual querying around.

joshuatf reviewed (commented):
I'm thinking more and more this should be handled in third party plugins.

We could offer an argument to force filters off in the `get_notes()` call, but I think this is not in the spirit of filters.

Do we have examples of the filters/clauses in question that are causing performance issues? cc @louwie17

jacob-sewell committed:
Add get_notes_with_type( $note_type ) method to Automattic\WooCommerce\Admin\Notes.
Name selected for consistency with preexisting get_notes_with_name( $name ) method.

jacob-sewell committed:
Use new get_notes_with_type() in Notes::possibly_delete_marketing_notes().

jacob-sewell committed:
Use new get_notes_with_type() in Notes::possibly_delete_survey_notes().

joshuatf reviewed (commented):
Thanks for the continued effort, Jacob!

I'm still not 100% convinced that we need to rethink our filters a bit.  Do we have the use cases 3PDs are using to query notes?

My gut feeling is that we should remove the `where` clause filter entirely and promote usage of the API directly or filtering at the client level.  Allowing filtering of all notes leads to side effects and performance issues as we see here.

jacob-sewell commented on src/Notes/DataStore.php:
```diff
@@ -471,4 +471,20 @@ public function get_notes_with_name( $name ) {
+	 * @param string $note_type Type to search for.
+	 * @return array An array of matching note ids.
+	 */
+	public function get_notes_with_type( $note_type ) {
```
I'm happy to change the name despite the inconsistency.

jacob-sewell commented on src/Notes/Notes.php:
```diff
@@ -273,15 +269,10 @@ public static function possibly_delete_marketing_notes( $old_value, $value ) {
-				'status' => array( 'actioned' ),
-			)
-		);
+		$note_ids   = $data_store->get_notes_with_type( Note::E_WC_ADMIN_NOTE_MARKETING );
```
Probably problematic, since this one is a hard delete. The old version wouldn't have deleted rows with an unactioned status, and this one would delete all of them. I'm not sure off the top of my head why we'd want to hard delete only actioned notes, but that's definitely a change.

jacob-sewell committed:
Correct copy-pasta error querying for MARKETING notes instead of SURVEY notes in possibly_delete_survey_notes().

jacob-sewell committed:
Only delete actioned notes in possibly_delete_survey_notes(), which was the previous behavior.

jacob-sewell committed:
Changed get_notes_with_type() to get_note_ids_by_type() in Notes data store.

jacob-sewell committed:
Changelog for 7987/8206.

jacob-sewell commented:
I've put in a changelog and consider this ready to go because even though there are still other questions about how the `woocommerce_note_where_clauses` filter is used, I think they belong in their own issue, and the ability to get note ids by just type is a useful addition.

jacob-sewell committed:
Add missing method name in changelog for 7987/8206.

joelclimbsthings commented on src/Notes/Notes.php:
```diff
@@ -273,16 +269,11 @@ public static function possibly_delete_marketing_notes( $old_value, $value ) {
-			if ( $note ) {
+		foreach ( $note_ids as $note_id ) {
+			$note = self::get_note( $note_id );
+			if ( $note && ( $note->get_status() === Note::E_WC_ADMIN_NOTE_ACTIONED ) ) {
```
Do we need these inner parenthesis?

jacob-sewell replied:
We don't, but I had to look that up just now, which goes to readability. 🤷

joelclimbsthings reviewed (approved):
Thanks for this @jacob-sewell . I'll admit I'm not _entirely_ clear on the best way to test this, but I checked out the branch and gave it a spin (including disabling marketing), and all was well. Left one minor comment. Pre-approving 🚢

jacob-sewell committed:
Merge branch 'main' into fix/7987-avoid-get_notes-in-possibly-delete-survey-note
